crates/shared/src/signature_validator.rs

```diff
@@ -111,43 +107,48 @@ impl SignatureValidating for Web3SignatureValidator {
     ) -> Result<u64, SignatureValidationError> {
         let instance = ERC1271SignatureValidator::at(&self.web3, check.signer);
-        let is_valid_result = instance
-            .is_valid_signature(Bytes(check.hash), Bytes(check.signature.clone()))
-            .call()
-            .await;
+        let check = instance.is_valid_signature(Bytes(check.hash), Bytes(check.signature.clone()));
 
-        let is_valid_gas_estimate_with_tx_initiation = instance
-            .is_valid_signature(Bytes(check.hash), Bytes(check.signature))
-            .m
-            .tx
-            .estimate_gas()
-            .await
-            .map_err(SignatureValidationError::Execution)?;
-
-        // Since all gas amounts should be smaller the the blocksize of 15M,
-        // the following operation should never panic
-        let is_valid_gas_estimate = is_valid_gas_estimate_with_tx_initiation.as_u64()
-            - TRANSACTION_INITIALIZATION_GAS_AMOUNT;
-
-        parse_is_valid_signature_result(is_valid_result).map(|_| is_valid_gas_estimate)
+        let (result, gas_estimate) =
+            futures::join!(check.clone().call(), check.m.tx.estimate_gas());
+
+        check_erc1271_result(result?)?;
+
+        // Adjust the estimate we receive by the fixed transaction gas cost.
+        // This is because this cost is not payed by an internal call, but by
+        // the root transaction only.
+        Ok(gas_estimate?.as_u64() - TRANSACTION_INITIALIZATION_GAS_AMOUNT)
     }
 }
 
 /// The Magical value as defined by EIP-1271
 const MAGICAL_VALUE: [u8; 4] = hex!("1626ba7e");
 
-fn parse_is_valid_signature_result(
-    result: Result<Bytes<[u8; 4]>, MethodError>,
-) -> Result<(), SignatureValidationError> {
-    match result {
-        Ok(Bytes(value)) if value == MAGICAL_VALUE => Ok(()),
-        Ok(_) => Err(SignatureValidationError::Invalid),
+fn check_erc1271_result(result: Bytes<[u8; 4]>) -> Result<(), SignatureValidationError> {
+    if result.0 == MAGICAL_VALUE {
+        Ok(())
+    } else {
+        Err(SignatureValidationError::Invalid)
+    }
+}
+
+impl From<MethodError> for SignatureValidationError {
+    fn from(err: MethodError) -> Self {
         // Classify "contract" errors as invalid signatures instead of node
         // errors (which may be temporary). This can happen if there is ABI
         // compability issues or calling an EOA instead of a SC.
-        Err(err) if EthcontractErrorType::classify(&err) == EthcontractErrorType::Contract => {
-            Err(SignatureValidationError::Invalid)
+        match EthcontractErrorType::classify(&err) {
+            EthcontractErrorType::Contract => Self::Invalid,
+            _ => Self::Other(err.into()),
+        }
+    }
+}
+
+impl From<ExecutionError> for SignatureValidationError {
+    fn from(err: ExecutionError) -> Self {
+        match EthcontractErrorType::classify(&err) {
+            EthcontractErrorType::Contract => Self::Invalid,
+            _ => Self::Other(err.into()),
```

Comment:
Is this the additional error classification that turns the 500 error into a 400 eventually?
I always get a little lost when there are so many error conversions happening. 😅

Reply:
Our `Other` errors always become 500 errors. The `Invalid` gets turned into an `403 Unauthorised` error.

> I always get a little lost when there are so many error conversions happening. 😅

Hmm, maybe this is something we can refactor at a higher level, so its easier for us to reason about error response codes. There is this `IntoWarpReply` that is largely responsible for this, but maybe this needs to be changed for additional clarity.